pkg/controller/pipeline.go

```diff
@@ -0,0 +1,523 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"slices"
+	"strings"
+
+	"github.com/go-logr/logr"
+	configv1alpha1 "github.com/rigdev/rig/pkg/api/config/v1alpha1"
+	"github.com/rigdev/rig/pkg/api/v1alpha2"
+	"github.com/rigdev/rig/pkg/errors"
+	"golang.org/x/exp/maps"
+	v1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	LabelOwnedByCapsule = "rig.dev/owned-by-capsule"
+)
+
+func GetNew[T client.Object](req Request, key ObjectKey) T {
+	var t T
+	o := req.GetNew(key)
+	if c, ok := o.(T); ok {
+		t = c
+	}
+	return t
+}
+
+func GetCurrent[T client.Object](req Request, key ObjectKey) T {
+	var t T
+	o := req.GetCurrent(key)
+	if c, ok := o.(T); ok {
+		t = c
+	}
+	return t
+}
+
+func Get[T interface {
+	client.Object
+	comparable
+}](req Request, key ObjectKey,
+) T {
+	var t T
+	if o := GetNew[T](req, key); o != t {
+		return o
+	}
+	if o := GetCurrent[T](req, key); o != t {
+		return o
+	}
+	return t
+}
+
+type Object struct {
+	Current client.Object
+	New     client.Object
+}
+
+type ObjectKey struct {
+	client.ObjectKey
+	schema.GroupVersionKind
+}
+
+func (ok ObjectKey) String() string {
+	return fmt.Sprintf("%s/%s", ok.GroupVersionKind.String(), ok.ObjectKey.String())
+}
+
+func (ok ObjectKey) MarshalLog() interface{} {
```

Comment:
Does this need to return `interface{}`? It seems the return value is statically known

Reply:
It's matching an interface to generate log messages, so not something we dictate